include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -281,10 +282,12 @@ class Matrix
      *
      * @param data  The device_matrix_data structure.
      * @param partition  The global row and column partition.
+     *
+     * @return the index_map induced by the partitions and the matrix structure
      */
-    void read_distributed(
+    index_map<local_index_type, global_index_type> read_distributed(
```

Comment:
we made no promises about interface-stability in `experimental`

Reply:
TBH, I don't consider that really an interface break. This has an effect only on very serious template meta programming, which I don't expect users to do. Other than that, current code will compile, it just discards the return value.

Reply:
There is actually an interface break here, but not due to the return value. Now this function expects a `shared_ptr<Partition>` instead of a raw pointer. But this is done in #1543 if you want to discuss that.

Reply:
Somehow the old comments I made dont appear anymore, so it is a bit annoying

Reply:
But this is indeed an unfortunate interface break. I think the version with `shared_ptr`  should be okay, but I think users actually depend on `read_distributed`, so it will break existing code.

Reply:
I think the index map needs to store the partition. I've looked again into the `map_to_local` kernels, and there I actually need *all* arrays a partition stores. So it wouldn't make sense to only copy the needed data out of the partition and into the index-map.